File: python/grade-school/grade_school.py
class School:
    def __init__(self):
        self.grades = {}
        self.added_ok = []

    def add_student(self, name, grade):
        if name in self.roster():
            self.added_ok.append(False)
        else:
            if grade in self.grades:
                self.grades[grade].append(name)
            else:
                self.grades[grade] = [name,]
            self.added_ok.append(True)

    def roster(self):
        names = []
        for grade in sorted(self.grades.keys()):
            names += sorted(self.grades[grade])
        return names

    def grade(self, grade_number):
        return [] if grade_number not in self.grades else sorted(self.grades[grade_number])

    def added(self):
        return self.added_ok

File: ruby/grade-school/grade_school.rb
class School
  def initialize
    @grades = Hash([])
  end

  def add(name, grade_num)
    return false if roster.include? name
    if @grades.has_key?(grade_num)
      @grades[grade_num].push(name)
    else
      @grades[grade_num] = [name]
    end
    true
  end

  def roster
    names = []
    @grades.keys.sort!.each do |g|
      names += @grades[g].sort
    end
    names
  end

  def grade(grade_num)
    @grades.has_key?(grade_num) ? @grades[grade_num].sort : []
  end
end
